### ISSUE TITLE
Include pairs bug

### ISSUE THREAD

**Issue by mcombs33** [Circuitscape/Circuitscape.jl#132]
Hello,

Thanks for the work on the new Circuitscape in Julia!

I am attempting to use the "include pairs" functionality in Julia (specifying to include paris within 1000m in this case) and the output resistance matrix appears to show that many of the pairs that should be included were not calculated, and some pairs that should not have been included were calculated. 

To confirm this, I ran circuitscape using the GUI app with my desired settings, which resulted in a correct resistance matrix (all pairs that should be calculated were, with no extras). Then I used the .ini file created by the GUI run to complete an identical run in Julia (only changing the output and log names), but the output resistance matrix had many missing pairs. 

I am attaching files below, all with .txt extension to support upload:
1) ascii layer used in run  
2) resistance file from the Julia run 
3) resistance file from the GUI run
4) include pairs distance matrix 
5) .ini file used to run julia 

Thanks for your help with this issue!
Matt

[aes_recl.txt](https://github.com/Circuitscape/Circuitscape.jl/files/2189833/aes_recl.txt)
[AES_1to1000_julia_resistances.txt](https://github.com/Circuitscape/Circuitscape.jl/files/2189820/AES_1to1000_julia_resistances.txt)
[AES_1to1000_resistances.txt](https://github.com/Circuitscape/Circuitscape.jl/files/2189821/AES_1to1000_resistances.txt)
[n196utm_dist_1to1000.txt](https://github.com/Circuitscape/Circuitscape.jl/files/2189822/n196utm_dist_1to1000.txt)
[AES_1to1000_julia.txt](https://github.com/Circuitscape/Circuitscape.jl/files/2189825/AES_1to1000_julia.txt)

**Comment by ranjanan** [comment]
Hi @mcombs33, sorry for the late reply. I see that you've used a point file as well which is in your INI (rats_n262_....txt). Would you mind attaching that too? Ideally I'm looking to reproduce your Circuitscape run on my computer and then looking into it from there. Thank you!

**Comment by ranjanan** [comment]
I wonder if this is related to #122

**Comment by mcombs33** [comment]
Here is the ascii for the sample points. Again, with .asc changed to .txt for uploading

This may be related to that #122 issue, but reading through, there doesn't seem to be a solution proposed. None of these samples should be on top of one another, with only one sample per pixel.

Thanks for the help.

[rats_n262_utm_ascii_3.txt](https://github.com/Circuitscape/Circuitscape.jl/files/2202957/rats_n262_utm_ascii_3.txt)

**Comment by rmarrotte** [comment]
Hi, this is still an issue. Points that are on top of each other seem to get a 0. But it seems that any pair that the first point has is just set to -1. You'll see in this example that point 1 has 7 pairs, but only the last one gets processed. Every other pair does get calculated though.
[test_run.zip](https://github.com/Circuitscape/Circuitscape.jl/files/2394434/test_run.zip)

**Comment by kearney-sp** [comment]
Has there been any progress on this issue? I am having the same problem (only with 'excluded' pairs). Runs fine in the GUI (4.0.5) with identified pairs excluded. When I export the .ini and run with Julia, it will run but output sets all pairs to -1.0. If I ignore the excluded pairs by setting 'use_included_pairs = False' it runs fine in Julia with a logical output, but this is not the desired analysis. Many thanks!

**Comment by ranjanan** [comment]
Hi @kearney-sp , I've picked this up again and I'm trying to fix it. Will provide an update in the next few days.

**Comment by kearney-sp** [comment]
Hi @ranjanan , great, thanks for the update! An update on my end: I was able to get the 'use_included_pairs = True' to work as expected on a test set with a .txt file of included pairs (headers 'mode' and 'include'), but it still would not work with a file of excluded pairs (headers 'mode' and 'exclude').

Additionally, when I attempted to process my full dataset using the same .ini setup with an 'include' list, I am getting strange results. First, the total number of pair solves seems arbitrary and is not what is expected. The process usually begins to run (I have tried multiple) and then at some point I get an error message, but the process continues to run. However, it is apparently no longer in parallel as the solving is serial and CPU use drops (may be related to Issue #165 ) and then, usually once it has finished the most recently scheduled pairs, it just hangs altogether and stops processing (it finishes solving a pair, and then nothing - no error message and no completion message, could be relate to Issue #122 ?)

Thanks a bunch for your work to convert this to Julia! Looks to be very promising.

I'm happy to provide more info if it is useful. Here is an example of an error message:

ERROR: On worker 2:
KeyError: key (133, 274) not found
getindex at .\dict.jl:478
f at C:\Users\spkearne\.julia\packages\Circuitscape\coCa3\src\core.jl:161
#81 at C:\Users\spkearne\.julia\packages\Circuitscape\coCa3\src\core.jl:225
#112 at C:\cygwin\home\Administrator\buildbot\worker\package_win64\build\usr\share\julia\stdlib\v1.0\Distributed\src\process_messages.jl:269
run_work_thunk at C:\cygwin\home\Administrator\buildbot\worker\package_win64\build\usr\share\julia\stdlib\v1.0\Distributed\src\process_messages.jl:56
macro expansion at C:\cygwin\home\Administrator\buildbot\worker\package_win64\build\usr\share\julia\stdlib\v1.0\Distributed\src\process_messages.jl:269 [inlined]
#111 at .\task.jl:259

**Comment by ranjanan** [comment]
@kearney-sp thanks for testing this out. Could you check out the fixes that I have pushed on the branch `RA/include`? You can do so by doing: `] add Circuitscape#RA/include` cc: @mcombs33 @rmarrotte

**Comment by kearney-sp** [comment]
Gave it a go and it read the maps (identified the total number of nodes) and then failed, presumably when attempting to compute the number of pairs, with the following error:

ERROR: UndefVarError: warn not defined
Stacktrace:
 [1] #read_polymap#35(::Int64, ::Bool, ::Function, ::Type, ::String, ::Circuitscape.RasterMeta) at C:\Users\spkearne\.julia\packages\Circuitscape\dJx9M\src\io.jl:175
 [2] read_polymap(::Type, ::String, ::Circuitscape.RasterMeta) at C:\Users\spkearne\.julia\packages\Circuitscape\dJx9M\src\io.jl:165
 [3] read_point_map(::Type, ::String, ::Circuitscape.RasterMeta) at C:\Users\spkearne\.julia\packages\Circuitscape\dJx9M\src\io.jl:196
 [4] load_raster_data(::Type, ::Type, ::Dict{String,String}) at C:\Users\spkearne\.julia\packages\Circuitscape\dJx9M\src\io.jl:425
 [5] raster_pairwise(::Type, ::Type, ::Dict{String,String}) at C:\Users\spkearne\.julia\packages\Circuitscape\dJx9M\src\raster\pairwise.jl:18
 [6] _compute(::Type, ::Type, ::Dict{String,String}) at C:\Users\spkearne\.julia\packages\Circuitscape\dJx9M\src\run.jl:41
 [7] macro expansion at .\util.jl:213 [inlined]
 [8] compute(::String) at C:\Users\spkearne\.julia\packages\Circuitscape\dJx9M\src\run.jl:30
 [9] top-level scope at none:0

**Comment by kearney-sp** [comment]
@ranjanan I tried it out and it appears to be working with with a .txt file of included pairs (headers 'mode' and 'include') on the full dataset. I will let you know if I encounter an error. It still does not appear to work with 'exclude', but this is not such a big deal since one can always use define an included pair instead.

Many thanks!

**Comment by ranjanan** [comment]
@kearney-sp could you please attach an example of it not working with `exclude` please? I just tested it on my end and it worked.

**Comment by ranjanan** [comment]
@rmarrotte would you be okay if I used test_run.zip as a test for this?

**Comment by kearney-sp** [comment]
Try files from this link here: https://www.dropbox.com/sh/4ff57vahma73q8p/AAAB6e2ppsq7Ti6_eAY4hjq4a?dl=0

 

Works with the ‘include_test.txt’, but does not work with the ‘exclude_test.txt’. (note that you will have to change the path names in the .ini file).

 

Let me know if there is a better way to share these files..

**Comment by ranjanan** [comment]
@kearney-sp I pushed a fix for `exclude` mode on the same branch. Could you check it now please?

**Comment by kearney-sp** [comment]
I ran : ] add Circuitscape#RA/include again and restarted Julia. Tried again and same result - all -1.0's. Did you get my test set to work on your end with 'exclude'? Perhaps there is an issue i have overlooked with how my data is setup...

**Comment by ranjanan** [comment]
Oh my apologies, could you restart julia and try?

**Comment by kearney-sp** [comment]
No luck. Tried restarting (and reloading the #RA/include) with the same results - all -1.0s.

**Comment by ranjanan** [comment]
Ok. I pushed what I thought would be a fix, but this could well be a new issue. Let me check.

**Comment by ranjanan** [comment]
@kearney-sp I just ran the older software on your problem with "exclude_pairs.txt" and it calculated the resistance for points (1,2) even though you said asked for it to be excluded on the file. The output looked like it ignored the exclude pairs list altogether. Could you tell me why this is desired behaviour? Thanks in advance.

**Comment by ranjanan** [comment]
Hi @kearney-sp, would you mind attaching the output of the GUI and the Julia version please? I suspect Circuitscape master (which is what I'm running) might be giving a different answer.

**Comment by ranjanan** [comment]
Actually I just ran it on my laptop and I have the answer: 

[out_resistances.txt](https://github.com/Circuitscape/Circuitscape.jl/files/2762713/out_resistances.txt)

Is this the same answer you are getting? My question is: when you have listed these pairs to be excluded from the computation, shouldn't there be -1s at those corresponding outputs? For example, as mentioned in `exclude_pairs.txt`, the first pair to be excluded is (1,2) whereas in this output there is a resistance value computed there. Isn't this undesired behaviour? 

Does my question now make sense?

**Comment by kearney-sp** [comment]
So I just re-ran with the GUI and checked the output and get identical results when using include_pairs.txt and exclude_pairs.txt (both identical to your result). The include_pairs result is as expected: only the listed pairs are output.

 

When using exclude_pairs.txt, I would expect that everything EXCEPT the list pairs would be included (rather than setting to -1 as you mentioned, they would be left out altogether as happens with include_pairs.txt, only the opposite pairs).

 

I thought I had this working properly before with the GUI, but perhaps I was mistaken. At the moment it seems that it also does not treat ‘exclude’ as expected. I will try again and see if I can get the expected result, but perhaps before I was in error and never got it working properly after all!

 

Again, this is not a big issue, since one can always just use ‘include’. However the guide for 4.0 suggests otherwise.

**Comment by ranjanan** [comment]
I am closing this for now, but please open another issue with expected output. Thank you! Fixed by #167

**Comment by wpeterman** [comment]
Using v.5.5.3+, I'm getting errors when using the 'include_pairs_file". It appears that the output matrix is 1 column and row smaller when run the pairs file. ZIP file with example files attached.
[test.zip](https://github.com/Circuitscape/Circuitscape.jl/files/3225706/test.zip)
